### DISCUSSION TITLE
2024年开源操作系统训练营项目二：ArceOS 宏内核项目成果总结

### DISCUSSION THREAD

**Discussion by Azure-stars** [Starry-OS/.github#2] [General]
请大家在这里填写自己目前的成果与感想，格式如下：

- 仓库链接：
- 最终 commit ID（用于帮助老师了解成果）：
- 目前进展说明（可附图）：
- 感想（可选）：

**Comment by dynamic-pigeon** [comment]
仓库链接：https://github.com/chy669086/futex

最终 commit ID: bdcd56277ccb04a2a4a4297eeb3996c50faf5f68

目前进展

除了 futex 优先级部分、定时唤醒和 shared 锁其他基本实现完毕。因为阶段二仓库我这里不能编译 linkme 的代码所以适配进二阶段 os 的时候是直接复制源码的。

![image](https://github.com/user-attachments/assets/af79072f-bdc1-4413-ac59-6a52ba7e1d09)

![image](https://github.com/user-attachments/assets/76ed6831-f138-4256-83cb-541cc1d11798)

![image](https://github.com/user-attachments/assets/9d2bbb17-dec6-43b1-8dd8-cc15620f54c1)

**Reply by Azure-stars** [reply]
good!

**Reply by Azure-stars** [reply]
这是对模块化的一个很好的尝试，如果有兴趣的话之后可以进一步维护，应该会是很有意义的工作！

**Comment by BattiestStone4** [comment]
仓库链接：
ArceOS基座：https://github.com/BattiestStone4/arceos/tree/monolithic
Starry-on-ArceOS：https://github.com/BattiestStone4/Starry-On-ArceOS
最终 commit ID：
[ArceOS基座：0698410](https://github.com/BattiestStone4/arceos/commit/069841000b53567202458da4afc443f4ee0eb1c2)
[Starry-on-ArceOS：923d965](https://github.com/BattiestStone4/Starry-On-ArceOS/commit/923d9651d2a6c38c0f0c6edfe450ff55fc0c23b8)
目前进展说明：
未实现的系统调用列表更新如下：
| syscall_id | syscall_name |
| ---------- | ------------ |
| 34         | mkdirat      |
| 40         | mount        |
| 56         | openat       |
| 80          | fstat          |
| 153        | times        |
| 160        | uname        |
| 214        | brk          |
| 215          | munmap          |
| 221        | exec         |
| 222          | mmap          |

59 pipe2
![image](https://github.com/user-attachments/assets/4b81eba8-7fce-49f0-bbb5-244c02f8fb4a)
220 clone (可以执行子进程）
![image](https://github.com/user-attachments/assets/67ae7922-485d-491d-8b67-e3f0fcc6ee9e)
260 wait4 (waitpid不会error）
![image](https://github.com/user-attachments/assets/07f2113e-2cde-4fbf-aa08-af7663feb329)
yield可以实现交互输出：
![image](https://github.com/user-attachments/assets/d6b6942d-26d5-4d4c-9965-c67f268cb139)

感想：
从去年初见开源操作系统训练营到今天完成训练营全部流程，一年的时间又过去了。
这一次开源操作系统训练营，我完成了二阶段的五个实验，这在当时看上去是一座不可逾越的大山，如今我也翻过来了。
在宏内核方向的阶段，通过为Starry-on-ArceOS实现系统调用，我了解到当前ArceOS在宏内核方面的进程设计上仍有繁琐的地方（如郑老师所说，process后期需要维护的字段太多），于是催生了Starry-on-ArceOS将宏内核扩展部分分离出来并实现功能的这样一个组件。
此外，对系统调用的实现也加深了我对操作系统里一些概念的理解。尤其是在clone系统调用上，对于一次调用，两次返回的概念我得到了实践与切身的体会。
总而言之，这一次训练营整个流程下来，我在操作系统以及系统软件的开发上有了实战经验。操作系统也从书本上枯燥的知识变为了可以真正实现并在qemu上运行的二进制软件，这一切都是很令人感叹并且很有成就感的事情。
希望我在操作系统这条路上，可以越走越远吧。

**Reply by Azure-stars** [reply]
欢迎继续探索！

**Comment by greatbridf** [comment]
## 仓库链接：
https://github.com/greatbridf/Starry-On-ArceOS
https://github.com/greatbridf/Starry

## 最终 commit ID（用于帮助老师了解成果）：
[8784871](https://github.com/greatbridf/Starry-On-ArceOS/commit/8784871acf06de371ba65d2fd7b492f814dae0f7)
[2cd4958](https://github.com/greatbridf/Starry/commit/2cd49581c5e7439b8e23c65c4ce07a88a5156990)

## 目前进展说明（可附图）：
修改了一部分的编译脚本，现在在mac上可以编译调试运行。附带homebrew上mac端用的交叉编译器，用于交叉编译内核测试样例，修改自其他大佬写的编译脚本（很长时间没维护了）。[homebrew-musl-cross](https://github.com/greatbridf/homebrew-musl-cross)

实现了`mkdir open openat close brk dup dup2 chdir getcwd getpid clone`几个系统调用。
<img width="1081" alt="image" src="https://github.com/user-attachments/assets/d7b2c0d0-941a-4267-b9bb-f6b03a57f812" />
<img width="1053" alt="image" src="https://github.com/user-attachments/assets/d7e622ab-a991-4367-a71a-c63a23c5fdcc" />


现在每个进程的资源都是使用`AxNamespace`实现的，实现不是很正确，每个进程的`namespace`都是从全局的那个复制过来的，但是好像`AxNamespace`里面用的`AxResource`使用的是共享的语义，所以并不是想要的效果，修改的话应该把这些东西全都放到`TaskExt`里会好一些。

`clone`实现了，但是`AddrSpace`克隆地址空间的部分是将实际页中的内容复制过来，并没有实现cow，如果实现cow的话，应该是在`Backend`的那个地方加一种mapping，但是受限于时间的考虑，没有做这个。`clone`中其他的那些`flags`都还没做，现在是按`CLONE_VM`，`CLONE_FILES`这些做的，和`fork`的语义相似。

<img width="1152" alt="image" src="https://github.com/user-attachments/assets/3b1de240-25fa-490c-a824-e102d821989d" />

##感想（可选）：
这次训练营感觉收获很大，看了arceos的组件化内核的架构，感觉很好，也比较容易维护，在实现过程当中也提高了自己的能力吧。觉得在starry这样一个项目上实现一个宏内核的想法很厉害，实际写起来也很顺手，很方便。

因为期末时间不太够所以写的非常粗糙，很多地方的实现都是很临时的，不是非常正确，也非常遗憾。但是在参加这个训练营的过程当中非常开心，希望以后还能参加到更多开发的过程当中。

**Reply by greatbridf** [reply]
在自己机子上测试的给的预编译的测例，用Starry-New里的build_img.sh的封装的镜像，然后在riscv64架构下跑的测试，由于时间有限，没有测试所有平台的测例。

**Reply by Azure-stars** [reply]
非常感谢您对 mac 等环境支持的工作，以及 namespace 这部分工作做的也很好。如果您有兴趣之后继续参与开发，可以联系我～

**Comment by yiming-gu** [comment]
仓库链接：https://github.com/yiming-gu/Starry-On-ArceOS

最终 commit ID：a44b1471a356a91de889c55be22f92e1a476668c

目前进展：
实现了mkdir open close brk dup dup2 chdir getcwd getppid clone mmap munmap wait4系统调用
![image](https://github.com/user-attachments/assets/be1de1a5-42a2-40c0-bd36-7f376e75b6d6)
![image](https://github.com/user-attachments/assets/69616fba-7049-4ed9-bca1-5d871b3afcf4)
![image](https://github.com/user-attachments/assets/318d46bc-db8b-458d-b1cb-72627721f456)
![image](https://github.com/user-attachments/assets/e887f5bb-48e9-4ef5-af3f-dfc1c6da78ae)
![image](https://github.com/user-attachments/assets/5e1e7f2d-c9c7-41f0-85de-3df0a4903bab)

感想：
这次训练营从不会rust和从来没学过操作系统的小白，一步一步跟着训练营的节奏走，从学习rust到做rcore再到arceos组件化操作系统，收获和成长了很多，虽然现在代码实现还是以功能实现为主，没有把组件化的抽象、隔离的思想应用进去，但是先完成后完美，后面也会逐步完善，继续学习操作系统。

**Reply by Azure-stars** [reply]
欢迎继续参与下一次训练营，进一步提升自我～

**Comment by hxingjie** [comment]
- 仓库链接：https://github.com/hxingjie/Starry-On-ArceOS

- 最终 commit ID：890d39999ce3615fc12db51199a7fc0d7177a6c7

- 目前进展说明：已实现的系统调用

  ```
  gettimeofday, getcwd, dup, duo3, chdir, mkdirat, openat, close, execve, uname, fstat，clone(只实现了复制当前进程)
  ```

- 感想：刚准备开始学习rust语言的时候，偶然间在知乎看到了陈老师的帖子，了解到原来有一个这么优秀的训练营。经过四个阶段的学习，有了很大的收获。一阶段的rust学习让我了解到了这门优秀的系统编程语言，二阶段让我将考研时学习到的操作系统知识有了具象化的认识，三四阶段让我了解到组件化操作系统这一方向。越往后面学习，越认识到自己的rust编程能力还很薄弱，对系统的认识还很浅显，在之后的时间要系统性的学习rust语言，更深入地学习操作系统的知识，希望下一次的训练营能够做的更好。

**Reply by Azure-stars** [reply]
下一次训练营应该也不会很远，欢迎继续参与～

**Comment by yjymosheng** [comment]
#  仓库链接: https://github.com/yjymosheng/Starry-On-ArceOS/tree/main

#  最终 commit ID: 71650c57f8ef9a64a6bdf274d9890b5c0c96642d

#  目前进展: 
已经完成的syscall 

![a](https://github.com/user-attachments/assets/835aaaf3-a3a4-4fc4-9214-f5c58e8af0f2)

没有完成的syscall 

![b](https://github.com/user-attachments/assets/7a21edfe-817f-4f42-84ed-fe0e488d6122)


#  感想 :  

这时我第二次参加os训练营,第一次的时候连rcore都没有完成,这一次由于对syscall的理解不够,相比其他同学浪费了很多时间在img的调试上.

可能下一次就能拿到优秀学员证书了也说不定?一次更比一次强嘛

虽然我的四阶段的学习可能差强人意,但是对我来说算是打开了操作系统的大门,希望明年的操作系统大赛上,能够有新的突破.

通过这次训练营，我进行了第一次对操作系统的尝试,与还对系统调用有了更加全面的认识。编写一个自己的操作系统不再只是纸上谈兵，而是一种有希望实现的技术。回顾整个过程，既有苦涩，也有喜悦，更重要的是，这让我对操作系统这条路充满了信心和期待。

希望未来，我能够在操作系统道路上走得更远，探索更多未知的可能性！

**Reply by Azure-stars** [reply]
能够看到一步步的进步对自己来说就是有意义的，欢迎继续参与操作系统的工作，下次训练营也欢迎继续参与～

**Comment by nusakom** [comment]
仓库链接：https://github.com/[nusakom/OPENcv](https://github.com/nusakom/OPENcv)
目前进展说明：New-Starry跑完了test___all__.py。下载opencv的官方测例，实习一小部分，还没有使用提供的测例进行测试。
![395868529-e80f4fd7-9be9-476e-ac08-58bba535f516](https://github.com/user-attachments/assets/c
![394598580-b71284ab-4c8c-4bfb-b789-92f95debeee9](https://github.com/user-attachments/assets/16d4bb6e-ac59-46df-a71c-f7a4366f2e09)
155f02
![微信图片_20241221234029](https://github.com/user-attachments/assets/a7fed155-9541-4ced-a1c6-fa72f81d1459)
d-493c-44e2-8d34-fcffd34100a7)
感想：第一次坚持到结营，感觉写实验作业和真正做项目的差距很大。就想小白一样，加上网络有点拉跨，简单的东西一直上传不了。
因为参与过一期，第一阶段其实还可以的。今年的rcore2024A有点变化很让人期待下一个版本的，第三阶段在结课的期间确实不够认真没有很好的完成全部任务。第四阶段选了相对其他几个方向一个比较简单的方向，对比之前盲目追求高难度的感觉好很多，虽然异步那些比opencv这些熟悉一点，但是在逐步实现的过程中懂得了python的适配一系列之前没有涉及到的方向。很可惜这一次没有在starry上面跑完opencv的测例，后期找空余时间把这个完成。从一个只知道free rtos是什么东西的小白到有一点点的理解操作系统的结构还是进步很多的，参加一年的训练营收获了很多，遇到了很多热心的助教和同学，感觉这个是更可贵的。
希望未来可以在自己的操作系统上面跑自己写的编译器!

**Reply by nusakom** [reply]
![395868529-e80f4fd7-9be9-476e-ac08-58bba535f516](https://github.com/user-attachments/assets/75292b5d-9014-424f-b98a-360122a122d5)
上传失败了，重新上传一个在Old-Starry实现的python测例截图

**Reply by Azure-stars** [reply]
期待之后你将编译器跑起来的那天！（其实之前 Starry 已经跑起来了最基本的 gcc 编译 c 程序，不过可能不够完善）
欢迎继续参与操作系统的工作！
关于今年的rCore2024A您有什么建议或者想法吗，也可以在这里提出。

**Comment by xingmin1** [comment]
#### 仓库链接： 
- https://github.com/xingmin1/Starry-On-ArceOS/tree/oscomp
- https://github.com/xingmin1/Starry/tree/monolithic
#### 最终 commit ID（用于帮助老师了解成果）： 
- [5cfe66c](https://github.com/xingmin1/Starry/commit/5cfe66c5f824095d4915f8e86883647a43cadf29)
- [82bf72a](https://github.com/xingmin1/Starry-On-ArceOS/commit/82bf72aa1659e7662444d4f14167bb16da57e056)
#### 目前进展说明（可附图）： 
![图片](https://github.com/user-attachments/assets/2907e480-4d41-4398-baab-b3bc4e182241)
未实现的系统调用

1. SYS_umount2
2. SYS_mount
3. SYS_brk
4. SYS_times

#### 感想（可选）：
我一开始了解到 rcore 是在学堂在线的上的操作系统课，它有两个版本，我选择了 rust 版的，我对 rust 比较感兴趣， 可能是喜欢新的东西吧。我在写 c 加加时，心里负担较大，焦虑每一步的性能消耗，或者说是每一步的语意不清晰。而 rust 相比起来，语意就清晰了许多，而且消耗大的操作都要显式指明。当然这也可能是我系统入门 rust ，但没有系统入门 c 加加，对我来说 系统入门 rust 要比 c 加加简单很多。

我一开始觉得做事需要先打好基础，所以在一二年级就一直在打基础，但我参加完这个训练营后发现，打好基础也需要做事。我发现我在写代码时，每一天都在学新的知识，比单纯的通过学知识学到的知识还多。同时我也感受到了在 OS 中无处不在的抽象，通常要解决一个问题不需要抽象，但要解决一类问题就需要抽象，因为抽象可以降低解决这一类问题的复杂度，但对于我这样的初学者来说，要看懂某一功能的实现可能更加困难了，因为它的解决办法就不止牵扯这一个问题了。所以，我觉得在作做os 之前，先知道这一些抽象的含义，可能当时不懂，没关系，在代码中遇到时能意识到，就能大大减轻读懂代码的难度，性价比非常高。在这一点上，第三阶段的 PPT 就发挥了巨大作用。

对我来说，结营后我依然会接着完成尚未完成任务，去增补os的代码，争取有一个好的结束。

**Reply by Azure-stars** [reply]
C++ 和 rust 各自的特点应该都比较明显，不过单纯说性能上应该没有高低的差距，不过可能是习惯不同吧
在学习知识的过程中动手实践写代码会有很大的帮助，因为你不知道自己看懂了没有，等到上手就会发现很多困难。比如第二阶段很早就讲过了 fork 等操作，但是在四阶段重新实现 clone 的时候就会遇到很多问题。
欢迎继续参与操作系统的工作！

**Comment by wang-jiahua** [comment]
仓库链接：[wang-jiahua/Starry-New](https://github.com/wang-jiahua/Starry-New)
目前进展说明：
在 Starry-New 上通过了 Python 测例
![image](https://github.com/user-attachments/assets/1abfca89-2fe8-4959-88c0-99c89d8a9156)
比赛测例还没有跑通，尝试了 x86 和 aarch64，另外尝试了官方测例
![image](https://github.com/user-attachments/assets/92de706a-0e79-4ca9-8005-952369528481)
感想：
感谢陈老师的介绍，感谢郑老师的指导，训练营对我的帮助非常大，之前只学过操作系统一些皮毛，没有机会用 Rust 亲手实践。训练营带我领略了操作系统之美，之后也希望尽量完成最后阶段的内容，增强动手能力。再次感谢！祝训练营越办越好！
